File: internal/limiter/tokenbucket.go
package limiter

import (
	"context"
	_ "embed"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
)

//go:embed tokenbucket.lua
var tokenBucketScript string

// TokenBucket refills a per-client allowance at a steady rate and lets a
// client spend it in bursts up to the bucket's capacity.
//
// # How it differs from the sliding window
//
// The sliding window answers "how many requests in the last N seconds?"
// exactly, by keeping one sorted-set entry per request. That is precise, and
// it costs memory proportional to traffic: a client doing 10k requests a
// minute keeps 10k members alive.
//
// A token bucket keeps two numbers per client — tokens and a timestamp —
// regardless of traffic, and it expresses a different intent: a sustained rate
// with an explicit burst allowance, rather than a hard ceiling over a trailing
// window. Pick it for high-volume clients, or when bursts are something to
// accommodate rather than to stop.
//
// # Why this one uses a script
//
// Refilling requires reading the stored timestamp, computing elapsed time and
// writing the result back. MULTI/EXEC cannot do that: commands inside a
// transaction are queued and cannot branch on each other's results. So the
// arithmetic runs inside Redis as a Lua script, which is atomic for the same
// reason the transaction is — Redis runs it to completion without interleaving
// anyone else's commands.
type TokenBucket struct {
	rdb    redis.Cmdable
	clock  Clock
	script *redis.Script
	// expiryGrace pads the TTL past the time a bucket needs to refill fully,
	// so a key cannot vanish while it still holds a deficit.
	expiryGrace time.Duration
}

// TokenBucketOption configures a TokenBucket.
type TokenBucketOption func(*TokenBucket)

// WithBucketClock sets the time source, for tests.
func WithBucketClock(c Clock) TokenBucketOption {
	return func(t *TokenBucket) { t.clock = c }
}

// NewTokenBucket returns a TokenBucket backed by rdb.
func NewTokenBucket(rdb redis.Cmdable, opts ...TokenBucketOption) *TokenBucket {
	t := &TokenBucket{
		rdb:         rdb,
		clock:       SystemClock{},
		script:      redis.NewScript(tokenBucketScript),
		expiryGrace: time.Second,
	}
	for _, opt := range opts {
		opt(t)
	}
	return t
}

// Allow implements Algorithm.
func (t *TokenBucket) Allow(ctx context.Context, key string, rule Rule, cost int64) (Decision, error) {
	if cost <= 0 {
		cost = 1
	}

	// Burst defaults to the limit, which makes an unconfigured bucket behave
	// like a plain rate cap rather than silently refusing everything.
	capacity := rule.Burst
	if capacity <= 0 {
		capacity = rule.Limit
	}

	// A bucket that starts full needs one full window to refill from empty, so
	// the TTL is sized to the refill time of its whole capacity.
	refill := time.Duration(float64(capacity) / float64(rule.Limit) * float64(rule.Window))
	ttl := refill + t.expiryGrace

	// Run prefers EVALSHA and falls back to EVAL when the script is not yet
	// cached, so a restarted or failed-over Redis recovers by itself.
	raw, err := t.script.Run(ctx, t.rdb, []string{key},
		t.clock.Now().UnixMicro(),
		rule.Limit,
		rule.Window.Microseconds(),
		capacity,
		cost,
		ttl.Milliseconds(),
	).Int64Slice()
	if err != nil {
		return Decision{}, fmt.Errorf("token bucket script: %w", err)
	}
	if len(raw) != 4 {
		return Decision{}, fmt.Errorf("token bucket script returned %d values, want 4", len(raw))
	}

	return Decision{
		Allowed:    raw[0] == 1,
		Limit:      capacity,
		Remaining:  raw[1],
		RetryAfter: time.Duration(raw[2]) * time.Microsecond,
		ResetAfter: time.Duration(raw[3]) * time.Microsecond,
	}, nil
}

File: internal/limiter/tokenbucket.lua
-- Token bucket, evaluated atomically inside Redis.
--
-- KEYS[1]  bucket key
-- ARGV[1]  now, in microseconds
-- ARGV[2]  limit: tokens refilled per window
-- ARGV[3]  window, in microseconds
-- ARGV[4]  capacity: the burst a client may accumulate
-- ARGV[5]  cost of this request
-- ARGV[6]  key TTL, in milliseconds
--
-- Returns {allowed, tokens_remaining, retry_after_micros, reset_after_micros}.
--
-- This is a script rather than a MULTI/EXEC block because the refill is a
-- read-modify-write: how many tokens to add depends on the stored timestamp,
-- and a transaction cannot branch on a value it has not read yet. Queued
-- commands in MULTI see no intermediate results, so the arithmetic has to
-- happen server-side. A script gives the same atomicity the sliding window
-- gets from its transaction.

local now      = tonumber(ARGV[1])
local limit    = tonumber(ARGV[2])
local window   = tonumber(ARGV[3])
local capacity = tonumber(ARGV[4])
local cost     = tonumber(ARGV[5])
local ttl      = tonumber(ARGV[6])

local rate = limit / window  -- tokens per microsecond

local state  = redis.call('HMGET', KEYS[1], 'tokens', 'ts')
local tokens = tonumber(state[1])
local ts     = tonumber(state[2])

if tokens == nil or ts == nil then
  -- An unseen client starts with a full bucket.
  tokens = capacity
  ts = now
end

-- Clamp elapsed at zero: clocks on separate limiter instances can disagree,
-- and a negative delta would hand out tokens that were never earned.
local elapsed = now - ts
if elapsed < 0 then elapsed = 0 end

tokens = math.min(capacity, tokens + elapsed * rate)

local allowed = 0
if tokens >= cost then
  tokens = tokens - cost
  allowed = 1
end

redis.call('HSET', KEYS[1], 'tokens', tokens, 'ts', now)
redis.call('PEXPIRE', KEYS[1], ttl)

local retry = 0
if allowed == 0 then
  retry = math.ceil((cost - tokens) / rate)
end

local reset = math.ceil((capacity - tokens) / rate)

-- Redis truncates floats returned from Lua, so every value is made an integer
-- deliberately rather than by accident.
return {allowed, math.floor(tokens), retry, reset}
